src/main/java/org/springframework/data/domain/Range.java

```diff
@@ -87,11 +206,230 @@ public boolean contains(T value) {
 
 		Assert.notNull(value, "Reference value must not be null!");
 
-		boolean greaterThanLowerBound = lowerBound == null ? true
-				: lowerInclusive ? lowerBound.compareTo(value) <= 0 : lowerBound.compareTo(value) < 0;
-		boolean lessThanUpperBound = upperBound == null ? true
-				: upperInclusive ? upperBound.compareTo(value) >= 0 : upperBound.compareTo(value) > 0;
+		boolean greaterThanLowerBound = lowerBound.getValue() //
+				.map(it -> lowerBound.isInclusive() ? it.compareTo(value) <= 0 : it.compareTo(value) < 0) //
+				.orElse(true);
+
+		boolean lessThanUpperBound = upperBound.getValue() //
+				.map(it -> upperBound.isInclusive() ? it.compareTo(value) >= 0 : it.compareTo(value) > 0) //
+				.orElse(true);
 
 		return greaterThanLowerBound && lessThanUpperBound;
 	}
+
+	/* 
+	 * (non-Javadoc)
+	 * @see java.lang.Object#toString()
+	 */
+	@Override
+	public String toString() {
+
+		String lower = this.lowerBound.getValue().map(Object::toString).map(it -> {
+
+			if (this.lowerBound.isInclusive()) {
+				return "[".concat(it);
+			}
+			return "(".concat(it);
+		}).orElse("unbounded");
+
+		String upper = this.upperBound.getValue().map(Object::toString).map(it -> {
+
+			if (this.upperBound.isInclusive()) {
+				return it.concat("]");
+			}
+			return it.concat(")");
+		}).orElse("unbounded");
+
+		return String.format("%s-%s", lower, upper);
+	}
+
+	/**
+	 * Value object representing a boundary. A boundary can either be {@link #unbounded() unbounded},
+	 * {@link #inclusive(Comparable) including its value} or {@link #exclusive(Comparable) its value}.
+	 * 
+	 * @author Mark Paluch
+	 * @since 2.0
+	 * @soundtrack Mohamed Ragab - Excelsior Sessions (March 2017)
+	 */
+	@Value
+	public static class Boundary<T extends Comparable<T>> {
+
+		@SuppressWarnings({ "rawtypes",
+				"unchecked" }) private static final Range.Boundary<?> UNBOUNDED = new Boundary(Optional.empty(), true);
+
+		private final Optional<T> value;
+		private final boolean inclusive;
```

Comment:
I'd like to stay with "inclusive" as`isInclusive()` reads more natural to me than `isIncluding()`.